kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/LlvmModuleSpecificationImpl.kt

```diff
@@ -28,10 +29,18 @@ internal abstract class LlvmModuleSpecificationBase(protected val cachedLibrarie
     override fun containsPackageFragment(packageFragment: IrPackageFragment): Boolean =
             packageFragment.konanLibrary.let { it == null || containsLibrary(it) }
 
-    override fun containsDeclaration(declaration: IrDeclaration): Boolean =
-            declaration.konanLibrary.let { it == null || containsLibrary(it) }
-                    // When producing per-file caches, some declarations might be generated by the stub generator.
-                    && declaration.getPackageFragment() !is IrExternalPackageFragment
+    private val containsCache = mutableMapOf<IrDeclaration, Boolean>()
+
+    override fun containsDeclaration(declaration: IrDeclaration): Boolean = containsCache.getOrPut(declaration) {
```

Comment:
What is the purpose of inlining of `declaration.konanLibrary` here?

Reply:
To memoize more of the calls, by inlining the konan library recursive calls here we can store the result for the top level class in addition to all nested classes. 

Overall this method showed up on the profiler as being very heavy due to the repeated recursive calls in KonanLibrary -- I opted to memoize the data in the llvm module specification class instead of trying to add it to the IrDeclaration or adding a global cache to konanLibrary in IrUtils.

Reply:
Got it. Please, add a comment here explaining this, and probably some short comment/reference to here from `declaration.konanLibrary` (in case someone decides to change it).

Reply:
@ddolovov I'm not sure that I understand your concern here. Is this more related to the size of the cache? 

With this change I've moved the recursive calls into the llvm contains declaration function so now that function should be called multiple times for `IrDeclarations`.

Reply:
> so now that function should be called multiple times for IrDeclarations.

You mean it could be called multiple times for the same instance of `IrDeclaration` (and therefore it makes sense to cache per-declaration), right?

Reply:
That's correct, especially for parent IrDeclarations they would be called repeatedly